src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -961,12 +961,38 @@ public override BoundNode VisitPassByCopy(BoundPassByCopy node)
 
         public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node)
         {
-            Debug.Assert(!IsConditionalState);
-            VisitRvalue(node.Expression);
-
             bool negated = node.Pattern.IsNegated(out var pattern);
             Debug.Assert(negated == node.IsNegated);
 
+            Debug.Assert(!IsConditionalState);
+            if (VisitPossibleConditionalAccess(node.Expression, out var stateWhenNotNull))
+            {
+                switch (GetTopLevelNullTest(pattern))
+                {
+                    case true:
+                        SetConditionalState(stateWhenNotNull, State);
+                        break;
+                    case false:
+                        SetConditionalState(State, stateWhenNotNull);
+                        break;
+                }
+            }
+            else if (IsConditionalState)
+            {
+                // Note that propagation of conditional state here is intentionally performed only on "simple" constant patterns, not on negated or binary patterns.
+                if (node.Pattern is BoundConstantPattern { ConstantValue: { IsBoolean: true, BooleanValue: var value } })
```

Comment:
>ConstantValue: { IsBoolean: true, BooleanValue: var value } [](start = 59, length = 59)

Why only bool patterns?

Reply:
I realized this was what the spec requires as written, and didn't feel interested in the scenarios that were missed. Opened dotnet/csharplang#4664 and changed the implementation here.